### DISCUSSION TITLE
How to disable, silence, freeze terminals literally?

### DISCUSSION THREAD

**Discussion by KiddoV** [jcubic/jquery.terminal#918] [Q&A]
I have some dynamic terminals on a page and user can choose to disable/enable the terminal. The terminals are connected to a backend COM port which will send the string to terminal in frontend whenever it receive data.

I have tried:
```ts
term.freeze(true);
term.disable(true);
```
... but later on when I call:
```ts
term.echo("HELLO!");
```
...the `terminal` will still display this string!!!

Is this expected behavior? Or those disable terminal methods is only for disable input prompt??
How do I disable, silence, freeze terminals literally then?
Thanks,

**Comment by jcubic** [comment]
So freeze is only to disable the input from user, and `enable`/`disable` is like `focus`/`blur`, but `focus` have logic to handle multiple terminals, it's more like switching between terminals when you have more than one.

If you want to disable `echo` you need to do this on your end:

```javascript
if (!term.frozen()) {
   term.echo('HELLO!');
}
```

**Reply by KiddoV** [reply]
Okay, thank you!

**Comment by tejas-ae** [comment]
**`freeze()` and `disable()` only affect user input — `echo()` always works regardless.** Here is the cleanest pattern for your use case:

**Option 1 — Guard on your end (recommended)**
```javascript
let silenced = false;

function silenceTerminal(term) {
    silenced = true;
    term.freeze(true);
}

function resumeTerminal(term) {
    silenced = false;
    term.freeze(false);
}

// In your backend data handler:
socket.on("data", (str) => {
    if (!silenced) term.echo(str);
});
```

**Option 2 — Wrap echo directly on the instance**
```javascript
const originalEcho = term.echo.bind(term);
term.echo = (...args) => {
    if (!term.frozen()) originalEcho(...args);
};
```
After `term.freeze(true)`, any `term.echo()` call is silently dropped. Call `term.freeze(false)` to resume both input and output together.

The key distinction in the API:
- `freeze(true)` — disables keyboard input from the user
- `disable()` / `enable()` — equivalent to blur/focus, switches active terminal when you have multiple
- Neither prevents programmatic `echo()` calls — that is by design
